### PR TITLE
Issue #12409: Inconsistent allowedAbbreviations

### PR DESCRIPTION
**Pull request by arinmodi**
Closes : #12409 : Inconsistent allowedAbbreviations when a method is contains an underscore

Diff Regression config: https://gist.githubusercontent.com/arinmodi/89cc612081d60d9f9b72e090df7db612/raw/7a96b0e42c45afecc39694493ede50add32fde77/config3.xml

### CODE REVIEW

**Comment by arinmodi** [comment]
Currently, Not Focusing on CI, Once Logic is approved, I will do that

**Comment by romani** [comment]
Preappoved

**Comment by arinmodi** [comment]
> Preappoved . Please extend example in xdoc with underscore

Do you mean adding example on the website ?

**Comment by nrmancuso** [comment]
@arinmodi please generate check regression report

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java [hunk 1]
```diff
@@ -297,6 +297,38 @@
...
+ *     &lt;property name="allowedAbbreviations" value="ORDER, OBSERVATION, UNDERSCORE, TEST"/&gt;
+ * &lt;/module&gt;
+ * </pre>
+ * <p>Example:</p>
+ * <pre>
+ * public class InputAbbreviationAsWordInNameType7 {
+ *    void getTEST() {
+ *    } // OK
+ *
+ *    void getORDER_OBSERVATION() {} // OK
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java:316]: Please remove All examples below this line

**Comment by arinmodi** [comment]
Analysis : 

If there is more Uppercase letters than specified length before "_", it is generating violation. 

Without "_" it is allowing +1 character cause of Camel Case. 

Is this expected behaviour ?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java [hunk 2]
```diff
@@ -297,6 +297,38 @@
...
  * </pre>
  * <p>
+ * To configure to check variables, enforce
+ * no abbreviations (essentially camel case) except for
+ * words like 'ORDERS', 'OBSERVATION', 'UNDERSCORE', 'TEST'.
+ * </p>
+ * <p>Configuration:</p>
+ * <pre>
+ * &lt;module name="AbbreviationAsWordInName"&gt;
+ *     &lt;property name="allowedAbbreviations" value="ORDER, OBSERVATION, UNDERSCORE, TEST"/&gt;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java:307]: Keep only ORDER and TEST

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java [hunk 3]
```diff
@@ -297,6 +297,38 @@
...
+ * &lt;/module&gt;
+ * </pre>
+ * <p>Example:</p>
+ * <pre>
+ * public class InputAbbreviationAsWordInNameType7 {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java:312]: Name should be `Test`

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    void getTEST_OBSERVATION() {} // ok
+
+    void getTEST_UNDERSCORE() {} // ok
+
+    void getORDER() {} // ok
+
+    void getOBSERVATION() {} // ok
+
+    void getORDER_UNDERSCORE() {} // ok
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java:27]: Please extend testing with cases where violation should be.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java [hunk 5]
```diff
@@ -297,6 +297,38 @@
...
+ *    void getTEST() {
+ *    } // OK
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java:314]: Remove method.

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java (hunk 1 above)
Comment (arinmodi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java (hunk 2 above)
Comment (arinmodi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java]: Done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java (hunk 4 above)
Comment (arinmodi) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java (hunk 5 above)
Comment (arinmodi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java:314]: Done

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java (hunk 5 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java:314]: Not done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+*/
+
+
+package com.puppycrawl.tools.checkstyle.checks.naming.abbreviationaswordinname;
+
+public class InputAbbreviationAsWordInNameType7 {
+    void getORDER_OBSERVATION() {} // ok
+
+    void getNONE_Test() {} // violation
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java:14]: Please restore all test code and add more for violation cases and make even more to weird cases.

Input should be full of test code.
User examples in xdoc should be compact to show concept of validation.

**Review by romani — CHANGES_REQUESTED** [review]
Items item

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java (hunk 4 above)
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java]: Not done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java (hunk 1 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java]: Not done

Comment (arinmodi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java]: Is it ok, now ?

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java (hunk 4 above)
Comment (arinmodi) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java:35]: Ok ?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java (hunk 5 above)
Comment (arinmodi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java]: Ok ?

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java (hunk 6 above)
Comment (arinmodi) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java:21]: Ok ?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java [hunk 7]
```diff
@@ -297,6 +297,23 @@
  * }
  * </pre>
  * <p>
+ * To configure to check variables, enforce
+ * no abbreviations (essentially camel case) except for
+ * words like 'ORDERS', 'OBSERVATION', 'UNDERSCORE', 'TEST'.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java:302]: Sync with code

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java [hunk 8]
```diff
@@ -297,6 +297,23 @@
...
+ * <p>Configuration:</p>
+ * <pre>
+ * &lt;module name="AbbreviationAsWordInName"&gt;
+ *     &lt;property name="allowedAbbreviations" value="ORDER, TEST"/&gt;
+ * &lt;/module&gt;
+ * </pre>
+ * <p>Example:</p>
+ * <pre>
+ * public class Test {
+ *    void getORDER_OBSERVATION() {} // violation, at most 4 consecutive capital letters allowed
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java:313]: Put TEST instead of observation

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+allowedAbbreviations = ORDER, OBSERVATION, UNDERSCORE, TEST
+
+
+*/
+
+
+package com.puppycrawl.tools.checkstyle.checks.naming.abbreviationaswordinname;
+
+public class InputAbbreviationAsWordInNameType7 {
+    private int HBCK_LOCK_PATH; // violation
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java:12]: All violations should have message

**Review by romani — CHANGES_REQUESTED** [review]
Items

**Comment by arinmodi** [comment]
> Please explain why this is violation https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/13139b6_2022152055/reports/diff/Hbase/index.html#A1
> 
> It not more than 4 capital letters.
> 
> Items

Because default allowed length is 3 and Without Underscore it is allowing 4 because last character is considered as the next word's first character which should be capital in camel case.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java (hunk 7 above)
Comment (arinmodi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java (hunk 8 above)
Comment (arinmodi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java]: Done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java (hunk 9 above)
Comment (arinmodi) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java]: Done

**Comment by romani** [comment]
We should not treat _ as part word.
If user decided to use _ it is definitely words separation.

Please add this case to test input and we need to extend documentation about _ support.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    // violation below 'Abb.* in name '.*' must contain no more than '4' consecutive cap.* let.*.'
+    Boolean[] BOOL_VALS = { false, true };
+
+    void getTEST() {
+    } // ok
+
+    void getORDER_OBSERVATION() {} // ok
+
+    // violation below 'Abb.* in name '.*' must contain no more than '4' consecutive cap.* let.*.'
+    void getNONE_Test() {}
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java:24]: Make violation message like:
`name 'NONE' must contain no more than '4' .*cap`

We need to see name.

Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java:24]: Make violation message like:
`name 'NONE' must contain no more than '4' .*cap`

We need to see name.

**Review by romani — CHANGES_REQUESTED** [review]
I

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java [hunk 11]
```diff
@@ -297,6 +297,23 @@
...
+ * <p>Configuration:</p>
+ * <pre>
+ * &lt;module name="AbbreviationAsWordInName"&gt;
+ *     &lt;property name="allowedAbbreviations" value="ORDER, TEST"/&gt;
+ * &lt;/module&gt;
+ * </pre>
+ * <p>Example:</p>
+ * <pre>
+ * public class Test {
+ *    void getORDER_TEST() {} // OK, ignored
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java:313]: `ok, both abbreviations are allowed `

**Comment by arinmodi** [comment]
> We should not treat _ as part of word. If user decided to use _ it is definitely words separation.
> 
> Please add this case to test input and we need to extend documentation about _ support.

```getNON_ETest()```, is this good example to add in test input ?

**Comment by arinmodi** [comment]
Hey I have doubt, 

Suppose allowedAbbreviationLength is 3, so before "_" how many Consecutive Capital Letters should be allowed ?  3 or 4

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java (hunk 10 above)
Comment (arinmodi) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java:24]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java (hunk 11 above)
Comment (arinmodi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java]: Done

**Comment by arinmodi** [comment]
@romani ?

**Comment by arinmodi** [comment]
@romani, can I get update on this ?

**Comment by arinmodi** [comment]
Github, generate report

**Comment by arinmodi** [comment]
Hey @romani, Updated The Code to generate violation whenever more than (allowed length + 1) found

**Comment by romani** [comment]
Please keep CI green.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java [hunk 12]
```diff
@@ -297,6 +298,23 @@
...
+ * <p>Configuration:</p>
+ * <pre>
+ * &lt;module name="AbbreviationAsWordInName"&gt;
+ *     &lt;property name="allowedAbbreviations" value="ORDER, TEST"/&gt;
+ * &lt;/module&gt;
+ * </pre>
+ * <p>Example:</p>
+ * <pre>
+ * public class Test {
+ *    void getORDER_TEST() {} // OK, both abbreviations are allowed
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java:314]: We have a rule to have all code examples to be same
And only change configs and location of violations.
Examples above has MAX_ALLOWED. Let's reuse it

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+allowedAbbreviations = ORDER, OBSERVATION, UNDERSCORE, TEST
+
+
+*/
+
+
+package com.puppycrawl.tools.checkstyle.checks.naming.abbreviationaswordinname;
+
+public class InputAbbreviationAsWordInNameType7 {
+    // violation below 'Abb.* in name 'HBCKL_LOCK_PATH' must contain no .* than '4' .*cap.* let.*.'
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java:12]: Use style like
name 'HBCKL_LOCK_PATH' must contain no more than '4' .*cap`
The less `.*` the better.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+AbbreviationAsWordInName
+allowedAbbreviations = ORDER, OBSERVATION, UNDERSCORE, TEST
+
+
+*/
+
+
+package com.puppycrawl.tools.checkstyle.checks.naming.abbreviationaswordinname;
+
+public class InputAbbreviationAsWordInNameType7 {
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java:11]: 7 -> Snakestyle

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
+/*
+AbbreviationAsWordInName
+allowedAbbreviations = ORDER, OBSERVATION, UNDERSCORE, TEST
+
+
+*/
+
+
+package com.puppycrawl.tools.checkstyle.checks.naming.abbreviationaswordinname;
+
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java:10]: Please add more tests code that will trigger violations on different situations

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java (hunk 12 above)
Comment (arinmodi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java]: Done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java (hunk 13 above)
Comment (arinmodi) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java]: Done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java (hunk 14 above)
Comment (arinmodi) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java]: Done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java (hunk 15 above)
Comment (arinmodi) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameType7.java]: Done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameTypeSnakeStyle.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+
+    // violation below '.* name 'HBCKL_LOCK_PATH' must contain no more than '4' .* cap.*.'
+    private int HBCKL_LOCK_PATH;
+
+    Boolean[] BOOL_VALS = { false, true };
+
+     // violation below '.* name '__DEMOS__TESTS_VAR' must contain no more than '4' .* cap.*.'
+    private int __DEMOS__TESTS_VAR = 5;
+
+    private int __DEMO__TEST_VAR = 6;
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameTypeSnakeStyle.java:23]: Please add line where violation is placed on last word after underscore

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameTypeSnakeStyle.java [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+public class InputAbbreviationAsWordInNameTypeSnakeStyle {
+    // violation below '.* name 'FLAG_IS_FIRST_RUN' must contain no more than '4' .* cap.*.'
+    public boolean FLAG_IS_FIRST_RUN = false;
+
+    // violation below '.* name 'HBCKL_LOCK_PATH' must contain no more than '4' .* cap.*.'
+    private int HBCKL_LOCK_PATH;
+
+    Boolean[] BOOL_VALS = { false, true };
+
+     // violation below '.* name '__DEMOS__TESTS_VAR' must contain no more than '4' .* cap.*.'
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameTypeSnakeStyle.java:20]: I pretty sure you can remove leading `.*`. Please do

Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameTypeSnakeStyle.java:20]: I pretty sure you can remove leading `.*`. Please do

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameTypeSnakeStyle.java (hunk 16 above)
Comment (arinmodi) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameTypeSnakeStyle.java:23]: Done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameTypeSnakeStyle.java (hunk 17 above)
Comment (arinmodi) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/InputAbbreviationAsWordInNameTypeSnakeStyle.java]: Done

**Comment by arinmodi** [comment]
Github, generate report

**Comment by arinmodi** [comment]
Github, generate report

**Comment by arinmodi** [comment]
@romani, is it good now ?

**Comment by romani** [comment]
A way better.
Please add test code with digits as hint us at https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/029cc0c_2022162445/reports/diff/Hbase/index.html#A1 to let us be aware that it is part of word or not.
Leading number, trailing numbers, numbers in mid of word.

**Comment by arinmodi** [comment]
> A way better. Please add test code with digits as hint us at https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/029cc0c_2022162445/reports/diff/Hbase/index.html#A1 to let us be aware that it is part of word or not. Leading number, trailing numbers, numbers in mid of word.

Done, Leading Numbers Not Possible

**Comment by romani** [comment]
Please add cases like TEST23456, 234VIO LATION ....

**Comment by arinmodi** [comment]
> 

I think java has rule that identifier must start with letter

**Review by romani — APPROVED** [review]
Good job.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java [hunk 18]
```diff
@@ -583,8 +605,17 @@ private String getDisallowedAbbreviation(String str) {
             else if (abbrStarted) {
                 abbrStarted = false;
 
-                final int endIndex = index - 1;
-                result = getAbbreviationIfIllegal(str, beginIndex, endIndex);
+                final int endIndex;
+                final int allowedLength;
+                if (symbol == '_') {
+                    endIndex = index;
+                    allowedLength = allowedAbbreviationLength + 1;
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java:612]: I am not fully following. Why do we need to pretend the property `allowedAbbreviationLength` is larger than it is? Why isn't the indexes enough?

Comment (arinmodi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java:612]: Cause we are passing the index of underscore, not last character that's why.

Comment (arinmodi) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java:612]: @rnveach, you can check previous commit and it's report for better understanding

**Review by rnveach — APPROVED** [review]

**Pull request merged**